File: ios/FamilyMovieNight/Models/Group.swift
import Foundation

struct Group: Codable, Identifiable, Equatable {
    let groupId: String
    let name: String
    let createdBy: String
    let streamingServices: [String]
    let memberCount: Int
    let createdAt: String
    let members: [GroupMember]

    var id: String { groupId }
}

struct GroupMember: Codable, Identifiable, Equatable {
    let userId: String
    let displayName: String
    let avatarKey: String
    let role: String
    let joinedAt: String
    // Slice C4/C5: Managed member fields — optional for backward compat with existing records
    let isManaged: Bool?
    let parentUserId: String?
    let memberType: String?  // "independent" | "managed"

    var id: String { userId }
    var isCreator: Bool { role == "creator" }
    var isManagedMember: Bool { isManaged == true }

    // Convenience init for code that predates managed member fields
    init(
        userId: String,
        displayName: String,
        avatarKey: String,
        role: String,
        joinedAt: String,
        isManaged: Bool? = nil,
        parentUserId: String? = nil,
        memberType: String? = nil
    ) {
        self.userId = userId
        self.displayName = displayName
        self.avatarKey = avatarKey
        self.role = role
        self.joinedAt = joinedAt
        self.isManaged = isManaged
        self.parentUserId = parentUserId
        self.memberType = memberType
    }
}

struct CreateGroupRequest: Codable {
    let name: String
}

struct InviteResponse: Codable {
    let inviteId: String
    let inviteToken: String
    let inviteUrl: String
    let status: String
    let expiresAt: String
}

struct AcceptInviteResponse: Codable {
    let groupId: String
    let groupName: String
    let role: String
}

struct CreateManagedMemberRequest: Codable {
    let displayName: String
    let avatarKey: String
}

struct CreateManagedMemberResponse: Codable {
    let userId: String
    let displayName: String
    let avatarKey: String
    let role: String
    let memberType: String
}

struct InviteListResponse: Codable {
    let invites: [InviteSummary]
}

struct InviteSummary: Codable, Identifiable {
    let inviteId: String
    let status: String
    let createdAt: String
    let expiresAt: String

    var id: String { inviteId }
}

File: backend/src/models/group.ts
import { z } from "zod";

export const CreateGroupSchema = z.object({
  name: z.string().min(1).max(40),
});

export const UpdateGroupSchema = z.object({
  name: z.string().min(1).max(40).optional(),
  streaming_services: z.array(z.string().min(1).max(50)).max(20).optional(),
});

export const GroupSchema = z.object({
  group_id: z.string(),
  name: z.string(),
  created_by: z.string(),
  streaming_services: z.array(z.string()).default([]),
  member_count: z.number().default(0),
  created_at: z.string(),
});

export const GroupMemberSchema = z.object({
  group_id: z.string(),
  user_id: z.string(),
  role: z.enum(["creator", "member"]),
  member_type: z.enum(["independent", "managed"]).optional().default("independent"),
  joined_at: z.string(),
});

export type Group = z.infer<typeof GroupSchema>;
export type GroupMember = z.infer<typeof GroupMemberSchema>;
export type CreateGroupInput = z.infer<typeof CreateGroupSchema>;
export type UpdateGroupInput = z.infer<typeof UpdateGroupSchema>;
